### PR TITLE
Add ONNX Hub Manifest and Generation Code

### PR DESCRIPTION
**Pull request by mhamilton723**
First stage of #455 (Onnx hub)

### CODE REVIEW

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            continue
+        data_matrix.append(list(row.find_all(re_table_val)))
+
+    if typepy.is_empty_sequence(data_matrix):
+        raise ValueError("data matrix is empty")
+
+    return pd.DataFrame(data_matrix, columns=headers)
+
+
+markdown_files = list(glob.glob(join("..", "**", "README.md"), recursive=True))[1:]
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:37]: Explicitly scan for the "vision" and "text" folder:

```suggestion
globDirs = [join("..", "vision"), join("..", "text")]
markdown_files = []
for dir in globDirs:
    markdown_files += glob.glob(join(dir, "**", "README.md"), recursive=True)
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    source_dir = split(row["source_file"])[0]
+    model_file = row[feild].contents[0].attrs['href']
+    rel_path = os.sep.join(join(source_dir, model_file).split(os.sep)[1:])
+    with open(join("..", rel_path), "rb") as f:
+        bytes = f.read()
+        sha256 = hashlib.sha256(bytes).hexdigest()
+    return {
+        feild: rel_path,
+        feild.replace("_path", "") + "_sha": sha256,
+        feild.replace("_path", "") + "_bytes": len(bytes)}
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:89]: Typo "feild".
```suggestion
def get_file_info(row, field):
    source_dir = split(row["source_file"])[0]
    model_file = row[field].contents[0].attrs["href"]
    rel_path = os.sep.join(join(source_dir, model_file).split(os.sep)[1:])
    with open(join("..", rel_path), "rb") as f:
        bytes = f.read()
        sha256 = hashlib.sha256(bytes).hexdigest()
    return {
        field: rel_path,
        field.replace("_path", "") + "_sha": sha256,
        field.replace("_path", "") + "_bytes": len(bytes),
    }
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:90]: Add a function to generate tags based on folder structure.
```suggestion
def get_model_tags(row):
    source_dir = split(row["source_file"])[0]
    raw_tags = source_dir.split(os.sep)[1:]
    return [tag.replace("_", " ") for tag in raw_tags]
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+output = []
+for i, row in renamed.iterrows():
+
+    model_info = get_file_info(row, "model_path")
+    model_path = model_info.pop("model_path")
+    metadata = model_info
+
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:98]:
```suggestion
    model_tags = get_model_tags(row)
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        print("malformed opset {} in {}".format(row["opset_version"].contents[0], row["source_file"]))
+        continue
+
+    if len(row["model"].contents) > 0:
+        output.append({
+            "model": row["model"].contents[0],
+            "model_path": model_path,
+            "onnx_version": row["onnx_version"].contents[0],
+            "opset_version": int(row["opset_version"].contents[0]),
+            "metadata": metadata})
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:117]:
```suggestion
            "metadata": metadata,
            "tags": model_tags,
        })
```

**Review by memoryz — CHANGES_REQUESTED** [review]
See comments.

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+
+renamed = df.rename(columns={col: prep_name(col) for col in df.columns.values})
+metadata_fields = [f for f in renamed.columns.values if f not in top_level_fields]
+
+
+def get_file_info(row, feild):
+    source_dir = split(row["source_file"])[0]
+    model_file = row[feild].contents[0].attrs['href']
+    rel_path = os.sep.join(join(source_dir, model_file).split(os.sep)[1:])
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:82]: Depending on the OS, the `rel_path` variable may use "\\" or "/" as the path separator. You may want to reformat it to use "/" before writing the output. This way, developer on different OS won't have to manually resolve the difference in path separator every time the manifest is regenerated.

```suggestion
    rel_path = "/".join(join(source_dir, model_file).split(os.sep)[1:])
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            "model": row["model"].contents[0],
+            "model_path": model_path,
+            "onnx_version": row["onnx_version"].contents[0],
+            "opset_version": int(row["opset_version"].contents[0]),
+            "metadata": metadata})
+    else:
+        print("Missing model in {}".format(row["source_file"]))
+
+with open(join("..", "ONNX_HUB_MANIFEST.json"), "w+") as f:
+    json.dump(output, f, indent=4)
```
Comment (memoryz) [inline workflow_scripts/update_onnx_hub_manifest.py:122]: Add top-level `repositories` list.
```suggestion
manifest_object = {
    "repositories": [
        {"name": "onnx/models/master", "type": "github", "owner": "onnx", "repo": "models", "ref": "master"}
    ],
    "models": output,
}

with open(join("..", "ONNX_HUB_MANIFEST.json"), "w+") as f:
    json.dump(manifest_object, f, indent=4)
```

File: workflow_scripts/update_onnx_hub_manifest.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        raise ValueError("data matrix is empty")
+
+    return pd.DataFrame(data_matrix, columns=headers)
+
+
+markdown_files = list(glob.glob(join("..", "**", "README.md"), recursive=True))[1:]
+manifest_object = []
+
+all_tables = []
+for markdown_file in markdown_files:
```
Comment (wenbingl) [inline workflow_scripts/update_onnx_hub_manifest.py:41]: We don't have any specification about how the ONNX models will be organized.
But there is a table in the https://github.com/onnx/models/blob/master/README.md which can be a starter to list all models.

File: workflow_scripts/update_onnx_hub_manifest.py (hunk 7 above)
Comment (mhamilton723) [reply workflow_scripts/update_onnx_hub_manifest.py:122]: > Add top-level `repositories` list.

One issue i forsee is the inclusion of the "master" in this as the client can point to any version of this by commit hash and branch name. I don't think the client will need any additional info about the repository but let me know what you have in mind for this info

File: workflow_scripts/update_onnx_hub_manifest.py (hunk 8 above)
Comment (mhamilton723) [reply workflow_scripts/update_onnx_hub_manifest.py:44]: I pulled out all info from inner tables to generate this manifest these are the only issues encountered:

Unrecognized table columns in file ../vision/body_analysis/age_gender/README.md: ['Model (Caffe)' 'Download' 'ONNX version' 'Opset version' 'Dataset']
Unrecognized table columns in file ../vision/object_detection_segmentation/duc/README.md: ['Filename' 'Size' 'Details']
no model_with_data in file ../vision/body_analysis/ultraface/README.md
no model_with_data in file ../vision/body_analysis/ultraface/README.md
no model_with_data in file ../vision/object_detection_segmentation/yolov2-coco/README.md
malformed opset ONNX 9, ONNX.ML 1 in ../text/machine_comprehension/bidirectional_attention_flow/README.md
no model_with_data in file ../text/machine_comprehension/t5/README.md
no model_with_data in file ../text/machine_comprehension/t5/README.md

Comment (wenbingl) [reply workflow_scripts/update_onnx_hub_manifest.py]: I just checked some links in table in the README are all alive, as well as the onnx model link.
Do you know what causes the issues you mentioned?

File: workflow_scripts/update_onnx_hub_manifest.py (hunk 7 above)
Comment (memoryz) [reply workflow_scripts/update_onnx_hub_manifest.py]: Discussed offline. Feel free to reject this change.

File: workflow_scripts/update_onnx_hub_manifest.py (hunk 8 above)
Comment (wenbingl) [reply workflow_scripts/update_onnx_hub_manifest.py]: If you decide to make dependence on the table in the top-level README, we probably can add a CI pipeline to valid any change made on this README will not break this generator script.

**Review by wenbingl — APPROVED** [review]

**Pull request merged**